packages/axes/src/types.ts

```diff
@@ -0,0 +1,122 @@
+import { SpringValues } from 'react-spring'
+import {
+    ScaleBand,
+    ScaleLinear,
+    ScaleOrdinal,
+    ScalePoint,
+    ScaleTime,
+    ScaleSymLog,
+    ScaleLogarithmic,
+} from 'd3-scale'
+import React from 'react'
+
+export type GridValuesBuilder<T> = T extends number
+    ? number[]
+    : T extends string
+    ? string[]
+    : T extends Date
+    ? Date[]
+    : never
+
+export type GridValues<T extends string | number | Date> = number | GridValuesBuilder<T>
+
+export type Point = {
+    x: number
+    y: number
+}
+
+export type ScaleWithBandwidth =
+    | (ScaleBand<any> & { type: 'band' })
+    | (ScalePoint<any> & { type: 'point' })
+
+export type AnyScale =
+    | (ScaleLinear<any, any> & { type: 'linear' })
+    | (ScaleOrdinal<any, any> & { type: 'ordinal' })
+    | (ScaleTime<any, any> & { useUTC: boolean; type: 'time' })
+    | (ScaleSymLog<any, any> & { type: 'symlog' })
+    | (ScaleLogarithmic<any, any> & { type: 'log' })
+    | ScaleWithBandwidth
+
+export type TicksSpec<Value extends string | number | Date> =
+    // exact number of ticks, please note that
+    // depending on the current range of values,
+    // you might not get this exact count
+    | number
+    // string is used for Date based scales,
+    // it can express a time interval,
+    // for example: every 2 weeks
+    | string
+    // override scale ticks with custom explicit values
+    | Value[]
+
+export type AxisLegendPosition = 'start' | 'middle' | 'end'
+
+export type ValueFormatter<Value extends string | number | Date> = (value: Value) => Value
+
+export interface AxisProp<Value extends string | number | Date> {
```

Comment:
`AxisProp` and `AxisProps` might be confusing as they both hold props, maybe we could use a more descriptive name?